File: frameworks/caddy/handler.go
package httparenahandler

import (
	"database/sql"
	"encoding/json"
	"fmt"
	"hash/crc32"
	"io"
	"math"
	"net/http"
	"os"
	"path/filepath"
	"runtime"
	"strconv"
	"strings"

	"github.com/caddyserver/caddy/v2"
	"github.com/caddyserver/caddy/v2/caddyconfig/caddyfile"
	"github.com/caddyserver/caddy/v2/caddyconfig/httpcaddyfile"
	"github.com/caddyserver/caddy/v2/modules/caddyhttp"
	_ "modernc.org/sqlite"
)

func init() {
	caddy.RegisterModule(Handler{})
	httpcaddyfile.RegisterHandlerDirective("httparena", parseCaddyfile)
}

type Rating struct {
	Score float64 `json:"score"`
	Count int     `json:"count"`
}

type ProcessedItem struct {
	ID       int      `json:"id"`
	Name     string   `json:"name"`
	Category string   `json:"category"`
	Price    float64  `json:"price"`
	Quantity int      `json:"quantity"`
	Active   bool     `json:"active"`
	Tags     []string `json:"tags"`
	Rating   Rating   `json:"rating"`
	Total    float64  `json:"total"`
}

type ProcessResponse struct {
	Items []ProcessedItem `json:"items"`
	Count int             `json:"count"`
}

type staticFile struct {
	data        []byte
	contentType string
}

type RawItem struct {
	ID       int      `json:"id"`
	Name     string   `json:"name"`
	Category string   `json:"category"`
	Price    float64  `json:"price"`
	Quantity int      `json:"quantity"`
	Active   bool     `json:"active"`
	Tags     []string `json:"tags"`
	Rating   Rating   `json:"rating"`
}

type Handler struct {
	dataset           []RawItem
	jsonLargeResponse []byte
	staticFiles       map[string]staticFile
	db                *sql.DB
}

func (Handler) CaddyModule() caddy.ModuleInfo {
	return caddy.ModuleInfo{
		ID:  "http.handlers.httparena",
		New: func() caddy.Module { return new(Handler) },
	}
}

func (h *Handler) Provision(ctx caddy.Context) error {
	path := os.Getenv("DATASET_PATH")
	if path == "" {
		path = "/data/dataset.json"
	}
	data, err := os.ReadFile(path)
	if err != nil {
		return nil // dataset not available, /json will 500
	}

	if err := json.Unmarshal(data, &h.dataset); err != nil {
		return nil
	}

	// Load large dataset for /compression
	largeData, err := os.ReadFile("/data/dataset-large.json")
	if err == nil {
		var largeDataset []struct {
			ID       int      `json:"id"`
			Name     string   `json:"name"`
			Category string   `json:"category"`
			Price    float64  `json:"price"`
			Quantity int      `json:"quantity"`
			Active   bool     `json:"active"`
			Tags     []string `json:"tags"`
			Rating   Rating   `json:"rating"`
		}
		if err := json.Unmarshal(largeData, &largeDataset); err == nil {
			largeItems := make([]ProcessedItem, len(largeDataset))
			for i, d := range largeDataset {
				largeItems[i] = ProcessedItem{
					ID: d.ID, Name: d.Name, Category: d.Category,
					Price: d.Price, Quantity: d.Quantity, Active: d.Active,
					Tags: d.Tags, Rating: d.Rating,
					Total: math.Round(d.Price*float64(d.Quantity)*100) / 100,
				}
			}
			h.jsonLargeResponse, _ = json.Marshal(ProcessResponse{Items: largeItems, Count: len(largeItems)})
		}
	}

	// Load static files
	mimeTypes := map[string]string{
		".css": "text/css", ".js": "application/javascript", ".html": "text/html",
		".woff2": "font/woff2", ".svg": "image/svg+xml", ".webp": "image/webp", ".json": "application/json",
	}
	h.staticFiles = make(map[string]staticFile)
	entries, err := os.ReadDir("/data/static")
	if err == nil {
		for _, e := range entries {
			if e.IsDir() {
				continue
			}
			d, err := os.ReadFile(filepath.Join("/data/static", e.Name()))
			if err != nil {
				continue
			}
			ext := filepath.Ext(e.Name())
			ct, ok := mimeTypes[ext]
			if !ok {
				ct = "application/octet-stream"
			}
			h.staticFiles[e.Name()] = staticFile{data: d, contentType: ct}
		}
	}

	// Open SQLite database
	db, err := sql.Open("sqlite", "file:/data/benchmark.db?mode=ro&immutable=1")
	if err == nil {
		db.SetMaxOpenConns(runtime.NumCPU())
		h.db = db
	}

	return nil
}

func sumQuery(r *http.Request) int64 {
	var sum int64
	for _, vals := range r.URL.Query() {
		for _, v := range vals {
			if n, err := strconv.ParseInt(v, 10, 64); err == nil {
				sum += n
			}
		}
	}
	return sum
}

func (h *Handler) ServeHTTP(w http.ResponseWriter, r *http.Request, next caddyhttp.Handler) error {
	path := r.URL.Path

	switch path {
	case "/pipeline":
		w.Header().Set("Content-Type", "text/plain")
		w.Header().Set("Server", "caddy")
		w.Write([]byte("ok"))
		return nil

	case "/json":
		if h.dataset != nil {
			items := make([]ProcessedItem, len(h.dataset))
			for i, d := range h.dataset {
				items[i] = ProcessedItem{
					ID: d.ID, Name: d.Name, Category: d.Category,
					Price: d.Price, Quantity: d.Quantity, Active: d.Active,
					Tags: d.Tags, Rating: d.Rating,
					Total: math.Round(d.Price*float64(d.Quantity)*100) / 100,
				}
			}
			resp, _ := json.Marshal(ProcessResponse{Items: items, Count: len(items)})
			w.Header().Set("Content-Type", "application/json")
			w.Header().Set("Server", "caddy")
			w.Header().Set("Content-Length", strconv.Itoa(len(resp)))
			w.Write(resp)
		} else {
			http.Error(w, "No dataset", 500)
		}
		return nil

	case "/baseline2":
		sum := sumQuery(r)
		w.Header().Set("Content-Type", "text/plain")
		w.Header().Set("Server", "caddy")
		fmt.Fprint(w, sum)
		return nil

	case "/baseline11":
		sum := sumQuery(r)
		if r.Method == "POST" && r.Body != nil {
			body, _ := io.ReadAll(r.Body)
			if n, err := strconv.ParseInt(string(body), 10, 64); err == nil {
				sum += n
			}
		}
		w.Header().Set("Content-Type", "text/plain")
		w.Header().Set("Server", "caddy")
		fmt.Fprint(w, sum)
		return nil

	case "/upload":
		if r.Method == "POST" && r.Body != nil {
			body, _ := io.ReadAll(r.Body)
			checksum := crc32.ChecksumIEEE(body)
			w.Header().Set("Content-Type", "text/plain")
			w.Header().Set("Server", "caddy")
			fmt.Fprintf(w, "%08x", checksum)
		} else {
			http.Error(w, "POST required", 405)
		}
		return nil

	case "/compression":
		if h.jsonLargeResponse != nil {
			w.Header().Set("Content-Type", "application/json")
			w.Header().Set("Server", "caddy")
			w.Write(h.jsonLargeResponse)
		} else {
			http.Error(w, "No dataset", 500)
		}
		return nil

	case "/db":
		if h.db == nil {
			http.Error(w, "DB not available", 500)
			return nil
		}
		minStr := r.URL.Query().Get("min")
		maxStr := r.URL.Query().Get("max")
		minPrice := 10.0
		maxPrice := 50.0
		if v, err := strconv.ParseFloat(minStr, 64); err == nil {
			minPrice = v
		}
		if v, err := strconv.ParseFloat(maxStr, 64); err == nil {
			maxPrice = v
		}
		rows, err := h.db.Query("SELECT id, name, category, price, quantity, active, tags, rating_score, rating_count FROM items WHERE price BETWEEN ? AND ? LIMIT 50", minPrice, maxPrice)
		if err != nil {
			http.Error(w, "Query failed", 500)
			return nil
		}
		defer rows.Close()
		var items []map[string]interface{}
		for rows.Next() {
			var dbId int
			var name, category, tags string
			var price float64
			var quantity int
			var active int
			var ratingScore float64
			var ratingCount int
			if err := rows.Scan(&dbId, &name, &category, &price, &quantity, &active, &tags, &ratingScore, &ratingCount); err != nil {
				continue
			}
			var tagsArr []string
			json.Unmarshal([]byte(tags), &tagsArr)
			items = append(items, map[string]interface{}{
				"id": dbId, "name": name, "category": category,
				"price": price, "quantity": quantity, "active": active == 1,
				"tags": tagsArr,
				"rating": map[string]interface{}{"score": ratingScore, "count": ratingCount},
			})
		}
		resp := map[string]interface{}{
			"items": items,
			"count": len(items),
		}
		w.Header().Set("Content-Type", "application/json")
		w.Header().Set("Server", "caddy")
		json.NewEncoder(w).Encode(resp)
		return nil
	}

	if strings.HasPrefix(path, "/static/") {
		name := path[8:]
		if sf, ok := h.staticFiles[name]; ok {
			w.Header().Set("Content-Type", sf.contentType)
			w.Header().Set("Server", "caddy")
			w.Header().Set("Content-Length", strconv.Itoa(len(sf.data)))
			w.Write(sf.data)
			return nil
		}
		http.NotFound(w, r)
		return nil
	}

	return next.ServeHTTP(w, r)
}

func (h *Handler) UnmarshalCaddyfile(d *caddyfile.Dispenser) error {
	return nil
}

func parseCaddyfile(h httpcaddyfile.Helper) (caddyhttp.MiddlewareHandler, error) {
	var handler Handler
	err := handler.UnmarshalCaddyfile(h.Dispenser)
	return &handler, err
}

var (
	_ caddyhttp.MiddlewareHandler = (*Handler)(nil)
	_ caddyfile.Unmarshaler       = (*Handler)(nil)
)

File: frameworks/nginx-openresty/handler.lua
local cjson = require "cjson"
local ffi = require "ffi"

local SQLITE_OPEN_READONLY = 1
local SQLITE_ROW = 100

local _M = {}

local json_resp = nil
local large_resp = ""
local static_files = {}
local sqlite = nil
local db_stmt_ptr = nil

function _M.init()
    local mime = {
        css = "text/css", js = "application/javascript", html = "text/html",
        woff2 = "font/woff2", svg = "image/svg+xml", webp = "image/webp",
        json = "application/json",
    }

    -- Load raw dataset for per-request processing
    local f = io.open(os.getenv("DATASET_PATH") or "/data/dataset.json", "r")
    if f then
        json_resp = cjson.decode(f:read("*a"))
        f:close()
    end

    -- Load large dataset
    f = io.open("/data/dataset-large.json", "r")
    if f then
        local data = cjson.decode(f:read("*a"))
        f:close()
        for _, item in ipairs(data) do
            item.total = math.floor(item.price * item.quantity * 100 + 0.5) / 100
        end
        large_resp = cjson.encode({items = data, count = #data})
    end

    -- Load static files
    local handle = io.popen("ls /data/static 2>/dev/null")
    if handle then
        for name in handle:lines() do
            local file = io.open("/data/static/" .. name, "rb")
            if file then
                local ext = name:match("%.(%w+)$")
                static_files[name] = {
                    data = file:read("*a"),
                    ct = mime[ext] or "application/octet-stream",
                }
                file:close()
            end
        end
        handle:close()
    end
end

function _M.init_worker()
    -- Load SQLite FFI in worker process (after fork)
    ffi.cdef[[
        typedef struct sqlite3 sqlite3;
        typedef struct sqlite3_stmt sqlite3_stmt;
        int sqlite3_open_v2(const char *filename, sqlite3 **ppDb, int flags, const char *zVfs);
        int sqlite3_close(sqlite3 *db);
        int sqlite3_exec(sqlite3 *db, const char *sql, void *cb, void *arg, char **errmsg);
        int sqlite3_prepare_v2(sqlite3 *db, const char *zSql, int nByte, sqlite3_stmt **ppStmt, const char **pzTail);
        int sqlite3_step(sqlite3_stmt *pStmt);
        int sqlite3_reset(sqlite3_stmt *pStmt);
        int sqlite3_bind_double(sqlite3_stmt *pStmt, int idx, double val);
        int sqlite3_column_int(sqlite3_stmt *pStmt, int iCol);
        double sqlite3_column_double(sqlite3_stmt *pStmt, int iCol);
        const char *sqlite3_column_text(sqlite3_stmt *pStmt, int iCol);
    ]]
    sqlite = ffi.load("libsqlite3.so.0")

    local db_handle = ffi.new("sqlite3*[1]")
    local rc = sqlite.sqlite3_open_v2("/data/benchmark.db", db_handle, SQLITE_OPEN_READONLY, nil)
    if rc == 0 and db_handle[0] ~= nil then
        local db_ptr = db_handle[0]
        sqlite.sqlite3_exec(db_ptr, "PRAGMA mmap_size=268435456", nil, nil, nil)
        local stmt_handle = ffi.new("sqlite3_stmt*[1]")
        rc = sqlite.sqlite3_prepare_v2(db_ptr,
            "SELECT id, name, category, price, quantity, active, tags, rating_score, rating_count FROM items WHERE price BETWEEN ? AND ? LIMIT 50",
            -1, stmt_handle, nil)
        if rc == 0 then
            db_stmt_ptr = stmt_handle[0]
        end
    end
end

local function sum_args()
    local args = ngx.req.get_uri_args()
    local sum = 0
    for _, v in pairs(args) do
        if type(v) == "table" then
            for _, val in ipairs(v) do
                sum = sum + (tonumber(val) or 0)
            end
        else
            sum = sum + (tonumber(v) or 0)
        end
    end
    return sum
end

function _M.pipeline()
    ngx.header["Content-Type"] = "text/plain"
    ngx.print("ok")
end

function _M.baseline11()
    local sum = sum_args()
    if ngx.req.get_method() == "POST" then
        ngx.req.read_body()
        local body = ngx.req.get_body_data()
        if body then
            sum = sum + (tonumber(body) or 0)
        end
    end
    ngx.header["Content-Type"] = "text/plain"
    ngx.print(string.format("%d", sum))
end

function _M.baseline2()
    ngx.header["Content-Type"] = "text/plain"
    ngx.print(string.format("%d", sum_args()))
end

function _M.json()
    if not json_resp then
        ngx.status = 500
        ngx.print("dataset not loaded")
        return
    end
    local items = {}
    for i, d in ipairs(json_resp) do
        local item = {}
        for k, v in pairs(d) do item[k] = v end
        item.total = math.floor(d.price * d.quantity * 100 + 0.5) / 100
        items[i] = item
    end
    ngx.header["Content-Type"] = "application/json"
    ngx.print(cjson.encode({items = items, count = #items}))
end

function _M.compression()
    if large_resp == "" then
        ngx.status = 500
        ngx.print("dataset not loaded")
        return
    end
    ngx.header["Content-Type"] = "application/json"
    ngx.print(large_resp)
end

function _M.upload()
    ngx.req.read_body()
    local body = ngx.req.get_body_data()
    if not body then
        local fname = ngx.req.get_body_file()
        if fname then
            local f = io.open(fname, "rb")
            if f then
                body = f:read("*a")
                f:close()
            end
        end
    end
    if not body then
        ngx.status = 400
        ngx.print("no body")
        return
    end
    ngx.header["Content-Type"] = "text/plain"
    ngx.print(string.format("%08x", ngx.crc32_long(body)))
end

function _M.static_file()
    local name = ngx.var.uri:match("/static/(.+)")
    if not name then
        ngx.status = 404
        ngx.print("not found")
        return
    end
    local sf = static_files[name]
    if not sf then
        ngx.status = 404
        ngx.print("not found")
        return
    end
    ngx.header["Content-Type"] = sf.ct
    ngx.print(sf.data)
end

function _M.db()
    if db_stmt_ptr == nil then
        ngx.status = 500
        ngx.print("DB not available")
        return
    end
    local args = ngx.req.get_uri_args()
    local min_price = tonumber(args.min) or 10.0
    local max_price = tonumber(args.max) or 50.0
    sqlite.sqlite3_bind_double(db_stmt_ptr, 1, min_price)
    sqlite.sqlite3_bind_double(db_stmt_ptr, 2, max_price)
    local items = {}
    while sqlite.sqlite3_step(db_stmt_ptr) == SQLITE_ROW do
        local id = sqlite.sqlite3_column_int(db_stmt_ptr, 0)
        local name = ffi.string(sqlite.sqlite3_column_text(db_stmt_ptr, 1))
        local category = ffi.string(sqlite.sqlite3_column_text(db_stmt_ptr, 2))
        local price = sqlite.sqlite3_column_double(db_stmt_ptr, 3)
        local quantity = sqlite.sqlite3_column_int(db_stmt_ptr, 4)
        local active = sqlite.sqlite3_column_int(db_stmt_ptr, 5) == 1
        local tags_str = ffi.string(sqlite.sqlite3_column_text(db_stmt_ptr, 6))
        local rating_score = sqlite.sqlite3_column_double(db_stmt_ptr, 7)
        local rating_count = sqlite.sqlite3_column_int(db_stmt_ptr, 8)
        items[#items + 1] = {
            id = id,
            name = name,
            category = category,
            price = price,
            quantity = quantity,
            active = active,
            tags = cjson.decode(tags_str),
            rating = { score = rating_score, count = rating_count },
        }
    end
    sqlite.sqlite3_reset(db_stmt_ptr)
    ngx.header["Content-Type"] = "application/json"
    ngx.print(cjson.encode({ items = items, count = #items }))
end

return _M
